socketcan_bridge/src/socketcan_recover_ctrl.cpp

```diff
@@ -0,0 +1,81 @@
+#include <socketcan_bridge/socketcan_recover_ctrl.h>
+
+namespace socketcan_bridge
+{
+    
+    /**
+     * @brief Constructor for the recovery control class
+     *
+     * Initializes the publisher, and sets up a timer to periodically check the bus state
+     */
+    SocketCANRecoverCtrl::SocketCANRecoverCtrl(ros::NodeHandle* nh, ros::NodeHandle* nh_param,
+        boost::shared_ptr<can::DriverInterface> driver) : driver_(driver)
+    {
+        state_pub_ = nh->advertise<can_msgs::CanState>("can_state", 1, true);
+
+        timer_ = nh->createWallTimer(ros::WallDuration(5), [this](const ros::WallTimerEvent& event) {recover();}, true, false);
+
+        state_listener_ = driver_->createStateListener(
+            can::StateInterface::StateDelegate(this, &SocketCANRecoverCtrl::stateCallback));
+
+    }
+
+    /**
+     * @brief Checks the state of the bus, if !statie.isReady() then the
+     * recover timer is started to fire in 5secs, otherwise we stop the timer
+     */
+    void SocketCANRecoverCtrl::stateCallback(const can::State & state) {
+        publishStatus(state);
+        if(!state.isReady())
+        {            
+            timer_.start();
+        }
+        else
+        {
+            timer_.stop();
```

Comment:
Why do you need to stop a one-shot timer?

Reply:
If status has been updated before the callback fires, we we're giving a little time for this system to "maybe" fix itself

Reply:
void ros::WallTimer::stop ()
  | Stop the timer. Once this call returns, no more callbacks will be called. Does nothing if the timer is already stopped.

ie, if status became ready we don't want the callback to fire

Reply:
There is no way that the state can change from bad to good without intervention (?).

Reply:
I wasn't sure if that statement would hold true in future versions of socketcan_interface